---
title: "Experiment set-up"
output: html_document
date: "2025-02-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Libraries
```{r}
library(xlsx)
library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(stringr)
library(glue)
library(viridis)
library(platetools)
set.seed(2025)
```




# 0. Set up
## 0.1. Set up directory
```{r}
working_from = "charite"

if (working_from == "home") {
  base_dir = "/home/alvaro/MyStuff/"
} else
  if (working_from == "charite") {
    base_dir = "C:/MyStuff/"
  }
```

## 0.2. Load data
```{r}
# Synthetases dataframes
synthetases <- as.data.frame(fread(paste(base_dir, "TSSC/Data/synthetases/trna_synthetases.csv", sep="")))

# Agar growth data
growth_agar <- as.data.frame(fread(paste(base_dir, "TSSC/Data/Boone_lab/Growth/Mine/Processed_by_them/growth_agar.tsv", sep="")))

# Master tRNA dataset
master_dataset <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Other/GtRNAdb/master_tRNA_dataset.csv", sep="")))

# Amino acid information dataset
amino_acids <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Other/GtRNAdb/amino_acids.csv", sep="")))

# Layout of samples in the 384-well plates we got from the Boone lab
sample_layout <- as.data.frame(fread(paste(base_dir, "TSSC/Data/Boone_lab/sample_layout.csv", sep="")))
```
--> Something that might be confusing about the sample layout: there is a column called "no growth after pinning", the strains that have "Yes" there are the strains that were not marked as sick by the Boone Lab when sent, but which did not grow when Wenxi pinned them here. 




# 1. Exploration 
## 1.1. Check how many tRNA-synthetase genes they have in their library
```{r}
# Grab the tRNA-synthetases that are in the library
trna_synths_in_library <- synthetases[synthetases$Gene.secondaryIdentifier %in% growth_agar$Sys.Name,]
trna_synths_in_library <- trna_synths_in_library[order(trna_synths_in_library$Gene.name),]

# Check, out of the tRNA genes that couldn't be KOd in the tRNA KO library, how many correspond to which amino acids - what defines this? I EXPLORE THIS FURTHER IN 1.11. IN THE MAIN TRNA .RMD FILE
sort(table(master_dataset$Isotype_from_anticodon[master_dataset$KOd == "No"]))
```

## 1.2. Check where these strains with TS alleles for tRNA synthetases are in the plates, and produce figures of the 384-well plates coloring by different things
By where the available strains with a TS allele for a tRNA synthetase are in each plate
```{r}
# For each 384-well plate I need to come up with a separate dataframe, with only the column with the appropriate well IDs, and another column I want to color the wells by
for (i in 1:4) {
  # Isolate strains in this plate and code how we want to color them
  temp <- sample_layout %>%
    filter(Plate384 == i) %>%
    mutate(Strain_type = case_when(Synthetase == "Yes" & !grepl("exclude", SB.notes, ignore.case = T) & Available == "Yes" ~ "Synthetase",
                              grepl("exclude", SB.notes, ignore.case = T) ~ "Exclude this well",
                              Synthetase == "Yes" & !grepl("exclude", SB.notes, ignore.case = T) & Available == "No" ~ "Synthetase - didn't grow",
                              #grepl("Wild type", Notes, ignore.case = T) ~ "BY4741",
                              #grepl("control added by Fang", Notes, ignore.case = T) ~ "BY4741 + p13623",
                              TRUE ~ "Other")) %>%
    dplyr::select(Strain_type, Well_ID_non_unique)
  
  # Create coloring scheme
  coloring_levels <- c("Other", "Synthetase", "Exclude this well", "Synthetase - didn't grow")    # CHANGE IT HERE IF I WANT TO COLOR CONTROL STRAINS OR SOMETHING ELSE!
  fixed_colors <- viridis(length(coloring_levels))
  names(fixed_colors) <- coloring_levels
  temp$Strain_type <- factor(temp$Strain_type, levels = coloring_levels)
  
  # Plot this plate
  raw_map(data = temp$Strain_type,
          well = temp$Well_ID_non_unique,
          plate = 384) +
    ggtitle(glue("Plate {i} - tRNA synthetases TS alleles")) +
    theme_dark() +
    scale_fill_manual(values = fixed_colors) +
    theme(axis.text=element_text(size=6)) +
    labs(fill = "Strain type")
  
  ggsave(paste(base_dir, "TSSC/Output/Plots/sample_layout/synthetases_plate_", i, ".png", sep=""))
}
```

## 1.3. Figure out which other strains to use for pilot
It would be nice to be able to use all the other strains around the those with tRNA-synth ts-alleles in plate 4, I will check if they are evenly distributed across the growth in 22ºC vs. growth in 26ºC plot.
That would be the strains in rows F and G of Plate 4! Let's have a look at them
```{r}
growth_agar <- growth_agar %>%
  mutate(possible_selection = case_when(Plate384 == 4 & (Row384 == which(letters == "f")|Row384 == which(letters == "g")) ~ TRUE,
                                        TRUE ~ FALSE))

ggplot(data = growth_agar, aes(x = Avg_AUC_22, y = Avg_AUC_26, col = as.factor(possible_selection))) +
  geom_point() +
  geom_abline() +
  labs(col = "Ratio 26ºC/22ºC") +
  theme_light() +
  xlab("AUC at 22ºC") +
  ylab("AUC at 26ºC")
```


## 1.4. Wenxi is back from his wedding (15.05.2025) - we will soon set up the pilot - I think we'll be using the 11 tRNA synthetase KO strains, and a few WT controls??? Let me subset the dataset to those and put the screenshot in the presentation for when we discuss this
```{r}
# We have 10 tRNA synthetase KOs, 1 WT, and we can add let's say 3 WTs+plasmid = 13 strains
strains_for_pilot <- sample_layout %>%
  filter(Sys.Name %in% synthetases$Gene.secondaryIdentifier | grepl("Wild type", Notes, ignore.case = T) | grepl("control added by Fang", Notes, ignore.case = T),
         Available == "Yes",
         !grepl("exclude", SB.notes, ignore.case = T)) %>%
  dplyr::select(-c(SB.notes))

# Actually let me make sure I only have 4 rows for the controls, one coming from each plate
strains_for_pilot_1 <- strains_for_pilot %>%
  filter(!grepl("control added", Notes, ignore.case = T))
strains_for_pilot_2 <- strains_for_pilot %>%
  filter(grepl("control added", Notes, ignore.case = T)) %>%
  distinct(Plate384, .keep_all = T) %>% 
  filter(Plate384 == 1 | Plate384 == 2 | Plate384 == 3)

strains_for_pilot_final <- rbind(strains_for_pilot_1, strains_for_pilot_2)

# Are there any other ts alleles that are related to tRNAs?
check <- sample_layout %>%
  filter(grepl("trna", Gene.name, ignore.case = T) & Available == "Yes")

# Add 3 more strains at random, just to make it 16, because 96 is a multiple of 16
rownames(strains_for_pilot_final) <- NULL

# Remove last column before sending it to Wenxi, I don't think it's useful at all and it could be confusing
strains_for_pilot_final <- strains_for_pilot_final %>%
  dplyr::select(-Well_ID_non_unique)
```


Decided to also add 10 lethal strains so that we can compare our tRNA synthetases to them - I'll try to pick lethal at different temperatures from the publication from Li et al., 2011
```{r}
# Load data
li_2011 <- as.data.frame(fread(paste(base_dir, "TSSC/Data/Articles/Li_et_al_2011/Li_et_al_S1.csv", sep = "")))

# Get available strains in our library for reference
avail <- sample_layout %>%
  filter(Available == "Yes")

# Check the distribution of lethal temperature
ggplot(data = li_2011, aes(x = lethal_temp_lower)) +
  geom_histogram() +
  theme_light()
sort(unique(li_2011$lethal_temp_lower))

# There are 5 (lower) lethal temperatures - I'll pick 2 strains from each to put in our experiment
lethal_26 <- li_2011 %>%
  filter(lethal_temp_lower == 26,
         Sys.Name %in% avail$Sys.Name)
lethal_26 <- lethal_26[sample(nrow(lethal_26), 2),]
lethal <- avail[avail$Sys.Name %in% lethal_26$Sys.Name,] %>%
  dplyr::distinct(Sys.Name, .keep_all = T)

lethal_30 <- li_2011 %>%
  filter(lethal_temp_lower == 30,
         Sys.Name %in% avail$Sys.Name)
lethal_30 <- lethal_30[sample(nrow(lethal_30), 2),]
temp <- avail[avail$Sys.Name %in% lethal_30$Sys.Name,]
lethal <- rbind(lethal, temp) %>%
  dplyr::distinct(Sys.Name, .keep_all = T)

lethal_35 <- li_2011 %>%
  filter(lethal_temp_lower == 35,
         Sys.Name %in% avail$Sys.Name)
lethal_35 <- lethal_35[sample(nrow(lethal_35), 2),]
temp <- avail[avail$Sys.Name %in% lethal_35$Sys.Name,]
lethal <- rbind(lethal, temp) %>%
  dplyr::distinct(Sys.Name, .keep_all = T)

lethal_37 <- li_2011 %>%
  filter(lethal_temp_lower == 37,
         Sys.Name %in% avail$Sys.Name)
lethal_37 <- lethal_37[sample(nrow(lethal_37), 2),]
temp <- avail[avail$Sys.Name %in% lethal_37$Sys.Name,]
lethal <- rbind(lethal, temp) %>%
  dplyr::distinct(Sys.Name, .keep_all = T)

lethal_385 <- li_2011 %>%
  filter(lethal_temp_lower == 38.5,
         Sys.Name %in% avail$Sys.Name)
lethal_385 <- lethal_385[sample(nrow(lethal_385), 2),]
temp <- avail[avail$Sys.Name %in% lethal_385$Sys.Name,]
lethal <- rbind(lethal, temp) %>%
  dplyr::distinct(Sys.Name, .keep_all = T)


# Join this to the previously selected strains
lethal <- lethal %>%
  dplyr::select(-c(SB.notes, Well_ID_non_unique)) %>%
  dplyr::mutate(Status = "Lethal")

strains_for_pilot_final <- strains_for_pilot_final %>%
  dplyr::mutate(Status = case_when(Synthetase == "Yes" ~ "Synthetase",
                                   Synthetase == "No" & Lab.ID == "Y12395" ~ "BY4741",
                                   Synthetase == "No" & Lab.ID != "Y12395" ~ "Control"))

strains_for_pilot_final_and_lethal <- rbind(strains_for_pilot_final, lethal)
```


```{r}
# Save as .xlsx to send it to Wenxi
write.xlsx(x = strains_for_pilot_final_and_lethal,
           file = paste(base_dir, "TSSC/Data/Pilot/setup/strains_for_pilot.xlsx", sep=""), 
           row.names = F)


# Plot the 4 plates coloring these strains to be used
for (i in 1:4) {
  # Isolate strains in this plate and code how we want to color them
  temp <- sample_layout %>%
    filter(Plate384 == i) %>%
    dplyr::mutate(Strain_type = case_when(Synthetase == "Yes" & (Well_ID %in% strains_for_pilot_final$Well_ID) ~ "Synthetase",
                                          Synthetase == "No" & (Well_ID %in% strains_for_pilot_final$Well_ID) & Lab.ID == "Y12395" ~ "BY4741",
                                          Synthetase == "No" & (Well_ID %in% strains_for_pilot_final$Well_ID) & Lab.ID != "Y12395" ~ "Control",
                                          Synthetase == "No" & (Well_ID %in% lethal$Well_ID) ~ "Lethal",
                                          TRUE ~ "Other")) %>%
    dplyr::select(Strain_type, Well_ID_non_unique) 
  
  # Create coloring scheme
  coloring_levels <- c("Other", "Synthetase", "BY4741", "Control", "Lethal")    # CHANGE IT HERE IF I WANT TO COLOR CONTROL STRAINS OR SOMETHING ELSE!
  fixed_colors <- viridis(length(coloring_levels))
  names(fixed_colors) <- coloring_levels
  temp$Strain_type <- factor(temp$Strain_type, levels = coloring_levels)
  
  # Plot this plate
  raw_map(data = temp$Strain_type,
          well = temp$Well_ID_non_unique,
          plate = 384) +
    ggtitle(glue("Plate {i} - tRNA synthetases TS alleles - strains for pilot")) +
    theme_dark() +
    scale_fill_manual(values = fixed_colors) +
    theme(axis.text=element_text(size=6)) +
    labs(fill = "Strain type")
  
  ggsave(paste(base_dir, "TSSC/Output/Plots/sample_layout/pilot/synthetases_plate_", i, "_strains_for_pilot.png", sep=""))
}
```



























