---
title: "Experiment set-up"
output: html_document
date: "2025-02-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Libraries
```{r}
library(xlsx)
library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(stringr)
library(glue)
library(viridis)
library(platetools)
```




# 0. Set up
## 0.1. Set up directory
```{r}
working_from = "charite"

if (working_from == "home") {
  base_dir = "/home/alvaro/MyStuff/"
} else
  if (working_from == "charite") {
    base_dir = "C:/MyStuff/"
  }

```

## 0.2. Load data
```{r}
# Synthetases dataframe
synthetases <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/trna_synthetases.csv", sep="")))

# Agar growth data
growth_agar <- as.data.frame(fread(paste(base_dir, "TSSC/Data/Boone_lab/Growth/Mine/Processed_by_them/growth_agar.tsv", sep="")))

# Master tRNA dataset
master_dataset <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Other/GtRNAdb/master_tRNA_dataset.csv", sep="")))

# Amino acid information dataset
amino_acids <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Other/GtRNAdb/amino_acids.csv", sep="")))

# Layout of samples in the 384-well plates we got from the Boone lab
sample_layout <- as.data.frame(fread(paste(base_dir, "TSSC/Data/Boone_lab/sample_layout.csv", sep="")))
```





# 1. Exploration 
## 1.1. Check how many tRNA-synthetase genes they have in their library
```{r}
# Grab the tRNA-synthetases that are in the library
trna_synths_in_library <- synthetases[synthetases$Gene.secondaryIdentifier %in% growth_agar$Sys.Name,]
trna_synths_in_library <- trna_synths_in_library[order(trna_synths_in_library$Gene.name),]

# Check, out of the tRNA genes that couldn't be KOd in the tRNA KO library, how many correspond to which amino acids - what defines this? I EXPLORE THIS FURTHER IN 1.11. IN THE MAIN TRNA .RMD FILE
sort(table(master_dataset$Isotype_from_anticodon[master_dataset$KOd == "No"]))
```

## 1.2. Check where these strains with TS alleles for tRNA synthetases are in the plates, and produce figures of the 384-well plates coloring by different things
By where the available strains with a TS allele for a tRNA synthetase are in each plate
```{r}
# For each 384-well plate I need to come up with a separate dataframe, with only the column with the appropriate well IDs, and another column I want to color the wells by
for (i in 1:4) {
  temp <- sample_layout %>%
    filter(Plate384 == i) %>%
    mutate(values = as.numeric(Synthetase == "Yes" & Available == "Yes")) %>%
    dplyr::select(values, Well_ID_non_unique)
  
  raw_map(data = temp$values,
          well = temp$Well_ID_non_unique,
          plate = 384) +
    ggtitle(glue("Plate {i} - tRNA synthetases TS alleles")) +
    theme_dark() +
    scale_fill_viridis() +
    theme(axis.text=element_text(size=6))
  
  ggsave(paste(base_dir, "TSSC/Output/Plots/sample_layout/synthetases_plate_", i, ".png", sep=""))
}
```

## 1.3. Figure out which other strains to use for pilot
It would be nice to be able to use all the other strains around the those with tRNA-synth ts-alleles in plate 4, I will check if they are evenly distributed across the growth in 22ºC vs. growth in 26ºC plot.
That would be the strains in rows F and G of Plate 4! Let's have a look at them
```{r}
growth_agar <- growth_agar %>%
  mutate(possible_selection = case_when(Plate384 == 4 & (Row384 == which(letters == "f")|Row384 == which(letters == "g")) ~ TRUE,
                                        TRUE ~ FALSE))

ggplot(data = growth_agar, aes(x = Avg_AUC_22, y = Avg_AUC_26, col = as.factor(possible_selection))) +
  geom_point() +
  geom_abline() +
  labs(col = "Ratio 26ºC/22ºC") +
  theme_light() +
  xlab("AUC at 22ºC") +
  ylab("AUC at 26ºC")
```

































