---
title: "initial_growth_curves_tRNA_synthetases_Wenxi"
output: html_document
date: "2025-03-06"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Libraries
```{r}
library(xlsx)
library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(stringr)
library(glue)
library(viridis)
library(platetools)
library(gridExtra)
library(roxygen2)
set.seed(2025)
```




# 0. Set up
## 0.1. Set up directory
```{r}
working_from = "charite"

if (working_from == "home") {
  base_dir = "/home/alvaro/MyStuff/"
} else
  if (working_from == "charite") {
    base_dir = "C:/MyStuff/"
  }


which_try = "second_try"                 # This indicates which of the 2 sets of OD measurements Wenxi did we want to look at 
if (which_try == "first_try") {
  try_for_plots = "First round"
} else
  if (which_try == "second_try") {
    try_for_plots = "Second round"
  }

```

## 0.2. Load data
```{r}
if (which_try == "first_try") {
  # First try - data from the first time Wenxi did the growth curves - only 24h
  growth_26 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/first_try/20250530-26C.xlsx", sep=""), sheetIndex = 1)
  growth_30 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/first_try/20250529-30C.xlsx", sep=""), sheetIndex = 1)
  growth_34 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/first_try/20250602-34C.xlsx", sep=""), sheetIndex = 1)
  growth_37 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/first_try/20250602-37C.xlsx", sep=""), sheetIndex = 1)
} else 
  if (which_try == "second_try") {
    # Second try - data from the second time Wenxi did the growth curves - 48h
    growth_22 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/second_try/20250623 TSGC-22-48H.xlsx", sep=""), sheetIndex = 1)
    growth_26 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/second_try/20250606-TSGC-26C-48.xlsx", sep=""), sheetIndex = 1)
    growth_30 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/second_try/20250623-TSGC-30C-48.xlsx", sep=""), sheetIndex = 1)
    growth_34 <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/second_try/20250606-TSGC-34C-48.xlsx", sep=""), sheetIndex = 1)
}

# Load helper functions
source(paste(base_dir, "tRNA_KOs/Code/R/Mine/0.general_use_functions.R", sep=""))

# Sample IDs (metadata to match sample names)
sample_ids <- read.xlsx(file = paste(base_dir, "TSSC/Data/pilot/pre_test_growth_curves/strains_for_pilot.xlsx", sep=""), sheetIndex = 1) %>%
  dplyr::mutate(Sample.ID = paste(Sys.Name, "_1", sep=""))

# Turn rows of the 384 well plates from numbers to letters, and create a column with the well IDs combining the two
sample_ids$Row384 <- sapply(sample_ids$Row384, function(i) LETTERS[i])
sample_ids <- add_well_IDs(df = sample_ids, 
                           column_numbers_col = "Col384", 
                           row_letters_col = "Row384")
colnames(sample_ids)[colnames(sample_ids) == "Well_ID"] <- "Well_ID_384"
```




# 1. Process growth data from PlateReader or whatever
## 1.1. Create functions to pre-process plate reader growth curves - TODO: NEED TO DOCUMENT THESE FUNCTIONS
```{r}
#' This function changes the sample names from the well ID to sample ID
#' 
function_to_set_new_colnames <- function(df, sample_ids) {
  old_colnames <- data.frame(colnames(df)[3:ncol(df)])
  colnames(old_colnames) <- c("GC96")
  temp_ids <- sample_ids %>%
    dplyr::select(GC96, Sample.ID)
  old_colnames <- left_join(old_colnames, temp_ids, by = "GC96")
  new_colnames <- c("Time", "Temperature", as.character(old_colnames$Sample.ID))
  
  colnames(df) <- new_colnames
  return(df)
}

#' Normalize the ODs for each sample to its OD at T = 0
#'
normalize_to_time_0 <- function(df) {
  keep_first_columns <- df %>%
    dplyr::select(Time, Temperature)
  only_data <- df %>%
    dplyr::select(!c(Time, Temperature))
  only_data <- as.data.frame(t(apply(as.matrix(only_data), 1, function(x) x/as.matrix(only_data)[1,])))
  
  out <- cbind(keep_first_columns, only_data)
  return(out)
}

#' Iterate over the strains, getting their averages and creating a new column for each of them
#' 
function_to_get_means_across_replicates <- function(df, strains) {
  # Come up with the average columns
  for (i in 1:length(strains)) {
    strain <- strains[i]
    df <- df %>%
      rowwise() %>%
      dplyr::mutate("{paste0(strain)}" := mean(c_across(contains(strain))), .after = Temperature)
  }
  
  # Remove all other columns
  df <- df %>%
    dplyr::select(Time, Temperature, any_of(strains))
  
  # Return the finished dataframe
  return(df)
}

#' Add a column with the temperature that they were supposed to have, then pivot them to long version
make_long_and_add_fixed_temp <- function(df, fixed_temp) {
  df <- df %>%
    dplyr::mutate(Temp_set = fixed_temp, .after = Temperature) %>%
    dplyr::rename(Temp_observed = Temperature) %>%
    pivot_longer(!c(Time, Temp_set, Temp_observed), names_to = "Strain", values_to = "OD_raw")
  return(df)
}

#' Create an Empty dataframe - contains "Empty" value for all entries, except Well ID column
#' which contains "Empty_n" with "n" being the number of the sample. We create 2 of these, 
#' one for each empty row in the plate, to add the information of those wells to the main 
#' data, so that we have all wells included there and don't get an error afterwards. Since
#' we want to have the same number of columns as in the main data, and this might change in 
#' time if I add extra columns there, I need to do this with this function, where I can 
#' provide n_cols = ncol(main_data)
#' The number of rows should always be 12 since that is the number of columns in a 96-well plate
create_empty_df <- function(prefix, n_cols, well_ID_numbers) {
  # First column: well IDs with given prefix (e.g., "A", "B", etc.)
  first_col <- paste(prefix, seq(1, 12), sep = "")
  
  # Remaining columns: all "Empty", repeated 12 times
  empty_matrix <- matrix("Empty", nrow = 12, ncol = n_cols - 2)
  
  # Last column: the one with the Well IDs - the IDs for this need to be provided, since they
  # must be different between one empty row and the next, otherwise we'll have duplicated well IDs
  last_col <- paste("Empty_", well_ID_numbers, sep="")
  
  # Combine first column with empty columns
  df <- data.frame(first_col, empty_matrix, last_col, stringsAsFactors = FALSE)
  
  return(df)
}




#' Wrapper function: takes the just imported data and returns the processed data, in either long or short format, normalized or not
#' 
#' @param df The data should be provided as a dataframe where the information of the machine and method used
#' are not included, only the OD measurements and the respective time and temperature
#' @param sample_ids Metadata about the samples, containing ...
#' @param replicates
#' @param empty_rows
#' @param fixed_temp
#'
process_plate_reader_growth_curves <- function(df, 
                                               sample_ids,
                                               replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                               empty_rows = c("D", "H"),
                                               fixed_temp, 
                                               normalized_to_t0, 
                                               long_version,
                                               wrong_plate) {
  
  # First of all we make sure we have the timepoints on the rows and the wells on the columns,
  # since one of the machines produces a dataframe in the opposite orientation
  if (sum(grepl("Cycle.Nr.", colnames(df))) == 1) {
    new_colnames <- df$Cycle.Nr.
    df <- df %>%
      dplyr::select(-Cycle.Nr.)
    df <- as.data.frame(t(df)) 
    colnames(df) <- new_colnames
  }
  
  # Set the first two colnames to be "Time" and "Temperature", also get rid of rownames
  colnames(df)[1] <- "Time"
  colnames(df)[2] <- "Temperature"
  rownames(df) <- NULL
  
  # Remove NAs in case full NA rows were introduced by Excel
  df <- na.omit(df)
  
  # There was a plate in which D row was not left empty (30ÂºC, first try) - need to do this horrible thing to take care of that
  if (wrong_plate == T) {
    ##   - row D in the plate is now row E in my dataset
    ##   - row E in the plate is now row F in my dataset
    ##   - row F in the plate is now row G in my dataset
    ##   - row G in the plate (empty) is now row D in my dataset
    ## SO REMEMBER THAT LABELS IN THE ANALYSIS FOR ROWS FROM D ARE CHANGED HERE!
    new_colnames <- c("Time", "Temperature",
                      paste("A", seq(1, 12, 1), sep = ""),
                      paste("B", seq(1, 12, 1), sep = ""), 
                      paste("C", seq(1, 12, 1), sep = ""), 
                      paste("E", seq(1, 12, 1), sep = ""), 
                      paste("F", seq(1, 12, 1), sep = ""),
                      paste("G", seq(1, 12, 1), sep = ""),
                      paste("D", seq(1, 12, 1), sep = ""),
                      paste("H", seq(1, 12, 1), sep = ""))
    colnames(df) <- new_colnames
  }
  
  # Extract the names of the rows that contain the replicates
  r1 <- replicates[[1]][1]
  r2 <- replicates[[1]][2]
  r3 <- replicates[[1]][3]
  r4 <- replicates[[2]][1]
  r5 <- replicates[[2]][2]
  r6 <- replicates[[2]][3]
  empty_1 <- empty_rows[1]
  empty_2 <- empty_rows[2]
  
  # Sample IDs - replicate it a couple times changing IDs, since the samples in wells A, B and C are triplicates of the same sample, and the same for E, F, G
  sample_ids_2 <- sample_ids %>%
    dplyr::mutate(GC96 = case_when(grepl(r1, GC96) ~ str_replace(GC96, r1, r2),
                                   grepl(r4, GC96) ~ str_replace(GC96, r4, r5)),
                  Sample.ID = paste(Sys.Name, "_2", sep=""))
  sample_ids_3 <- sample_ids %>%
    dplyr::mutate(GC96 = case_when(grepl(r1, GC96) ~ str_replace(GC96, r1, r3),
                                   grepl(r4, GC96) ~ str_replace(GC96, r4, r6)),
                  Sample.ID = paste(Sys.Name, "_3", sep=""))
  
  sample_ids_new <- rbind(rbind(sample_ids, sample_ids_2), sample_ids_3)
  
  # Need to add rows for the empty wells, otherwise I get an error later...
  d_sample_ids <- create_empty_df(prefix = empty_1,
                                  n_cols = ncol(sample_ids_new),
                                  well_ID_numbers = seq(1, 12, 1))
  colnames(d_sample_ids) <- colnames(sample_ids_new)
  
  h_sample_ids <- create_empty_df(prefix = empty_2,
                                  n_cols = ncol(sample_ids_new),
                                  well_ID_numbers = seq(13, 24, 1))
  colnames(h_sample_ids) <- colnames(sample_ids_new)
  
  sample_ids_new_2 <- rbind(rbind(sample_ids_new, d_sample_ids), h_sample_ids)
  
  
  # Set new column names, take averages of the 3 replicates, set to long format and put all temperatures together --> ready to plot!
  ## Set new colnames
  df <- function_to_set_new_colnames(df, sample_ids_new_2)
  
  ## Normalize to time 0
  df_norm <- normalize_to_time_0(df)
  
  ## Get mean across replicates
  strains <- unique(sample_ids_new_2$Sys.Name)
  df_norm_avg <- function_to_get_means_across_replicates(df_norm, strains)
  df_avg <- function_to_get_means_across_replicates(df, strains)
  
  ## Set to long format and add reference temperature column
  df_norm_avg_long <- make_long_and_add_fixed_temp(df_norm_avg, fixed_temp)
  df_avg_long <- make_long_and_add_fixed_temp(df_avg, fixed_temp)
  
  # Figure out the required type of output, and return it
  if (normalized_to_t0 == T & long_version == T) {
    df_norm_avg_long <- df_norm_avg_long %>%
      dplyr::rename(OD_norm = OD_raw)
    return(df_norm_avg_long)
  }
  else if (normalized_to_t0 == T & long_version == F) {
    df_norm_avg <- df_norm_avg %>%
      dplyr::rename(OD_norm = OD_raw)
    return(df_norm_avg)
  }
  else if (normalized_to_t0 == F & long_version == T) {
    return(df_avg_long)
  }
  else if (normalized_to_t0 == F & long_version == F) {
    return(df_avg)
  }
}







# Get full sample IDs for one plate (so I can plot the growth curve plates with raw_map())
# This happens inside the wrapper above, in fact this is just a part of it, but it is not 
# outputted from there nor would I like it to, this is just for a very punctual thing, at
# least for now
create_full_sample_ids_for_a_plate <- function(df, 
                                               sample_ids,
                                               replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                               empty_rows = c("D", "H"),
                                               fixed_temp, 
                                               normalized_to_t0, 
                                               long_version,
                                               wrong_plate) {
  
  # First of all we make sure we have the timepoints on the rows and the wells on the columns,
  # since one of the machines produces a dataframe in the opposite orientation
  if (sum(grepl("Cycle.Nr.", colnames(df))) == 1) {
    new_colnames <- df$Cycle.Nr.
    df <- df %>%
      dplyr::select(-Cycle.Nr.)
    df <- as.data.frame(t(df)) 
    colnames(df) <- new_colnames
  }
  
  # Set the first two colnames to be "Time" and "Temperature", also get rid of rownames
  colnames(df)[1] <- "Time"
  colnames(df)[2] <- "Temperature"
  rownames(df) <- NULL
  
  # Remove NAs in case full NA rows were introduced by Excel
  df <- na.omit(df)
  
  # There was a plate in which D row was not left empty (30ÂºC, first try) - need to do this horrible thing to take care of that
  if (wrong_plate == T) {
    ##   - row D in the plate is now row E in my dataset
    ##   - row E in the plate is now row F in my dataset
    ##   - row F in the plate is now row G in my dataset
    ##   - row G in the plate (empty) is now row D in my dataset
    ## SO REMEMBER THAT LABELS IN THE ANALYSIS FOR ROWS FROM D ARE CHANGED HERE!
    new_colnames <- c("Time", "Temperature",
                      paste("A", seq(1, 12, 1), sep = ""),
                      paste("B", seq(1, 12, 1), sep = ""), 
                      paste("C", seq(1, 12, 1), sep = ""), 
                      paste("E", seq(1, 12, 1), sep = ""), 
                      paste("F", seq(1, 12, 1), sep = ""),
                      paste("G", seq(1, 12, 1), sep = ""),
                      paste("D", seq(1, 12, 1), sep = ""),
                      paste("H", seq(1, 12, 1), sep = ""))
    colnames(df) <- new_colnames
  }
  
  # Extract the names of the rows that contain the replicates
  r1 <- replicates[[1]][1]
  r2 <- replicates[[1]][2]
  r3 <- replicates[[1]][3]
  r4 <- replicates[[2]][1]
  r5 <- replicates[[2]][2]
  r6 <- replicates[[2]][3]
  empty_1 <- empty_rows[1]
  empty_2 <- empty_rows[2]
  
  # Sample IDs - replicate it a couple times changing IDs, since the samples in wells A, B and C are triplicates of the same sample, and the same for E, F, G
  sample_ids_2 <- sample_ids %>%
    dplyr::mutate(GC96 = case_when(grepl(r1, GC96) ~ str_replace(GC96, r1, r2),
                                   grepl(r4, GC96) ~ str_replace(GC96, r4, r5)),
                  Sample.ID = paste(Sys.Name, "_2", sep=""))
  sample_ids_3 <- sample_ids %>%
    dplyr::mutate(GC96 = case_when(grepl(r1, GC96) ~ str_replace(GC96, r1, r3),
                                   grepl(r4, GC96) ~ str_replace(GC96, r4, r6)),
                  Sample.ID = paste(Sys.Name, "_3", sep=""))
  
  sample_ids_new <- rbind(rbind(sample_ids, sample_ids_2), sample_ids_3)
  
  # Need to add rows for the empty wells, otherwise I get an error later...
  d_sample_ids <- create_empty_df(prefix = empty_1,
                                  n_cols = ncol(sample_ids_new),
                                  well_ID_numbers = seq(1, 12, 1))
  colnames(d_sample_ids) <- colnames(sample_ids_new)
  
  h_sample_ids <- create_empty_df(prefix = empty_2,
                                  n_cols = ncol(sample_ids_new),
                                  well_ID_numbers = seq(13, 24, 1))
  colnames(h_sample_ids) <- colnames(sample_ids_new)
  
  sample_ids_new_2 <- rbind(rbind(sample_ids_new, d_sample_ids), h_sample_ids)
  
  
  return(sample_ids_new_2)
}
```

## 1.2. Run pre-processing - for the datasets from the first set of growth curves
Here I apply all the functions created above to the datasets from both the first and second set of growth curves done by Wenxi. I must have separate code for each of these 2 options since there are different temperatures in the 2 sets, and the 30ÂºC one in the first set has the wrong row empty. 
```{r}
# Those from the first set
if (which_try == "first_try") {
  # Here we get: 
  # - Long version
  # - Normalized to time 0
  # - Averaged across replicates
  growth_26_avg_norm_long <- process_plate_reader_growth_curves(df = growth_26,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 26,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_30_avg_norm_long <- process_plate_reader_growth_curves(df = growth_30,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 30,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = T)
  
  growth_34_avg_norm_long <- process_plate_reader_growth_curves(df = growth_34,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 34,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_37_avg_norm_long <- process_plate_reader_growth_curves(df = growth_37,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 37,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = F)
  # Join results from this into a single long dataframe with all temperatures, to plot from
  growth_long_full_norm <- rbind(rbind(rbind(growth_26_avg_norm_long, growth_30_avg_norm_long), growth_34_avg_norm_long), growth_37_avg_norm_long)
  
  
  
  # Here we get:
  # - Long version
  # - Not normalized
  # - Averaged across replicates
  growth_26_avg_long_raw <- process_plate_reader_growth_curves(df = growth_26,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 26,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_30_avg_long_raw <- process_plate_reader_growth_curves(df = growth_30,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 30,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = T)
  
  growth_34_avg_long_raw <- process_plate_reader_growth_curves(df = growth_34,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 34,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_37_avg_long_raw <- process_plate_reader_growth_curves(df = growth_37,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 37,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)
  # Join results from this into a single long dataframe with all temperatures, to plot from
  growth_long_full_raw <- rbind(rbind(rbind(growth_26_avg_long_raw, growth_30_avg_long_raw), growth_34_avg_long_raw), growth_37_avg_long_raw)
}




# Those from the second set
if (which_try == "second_try") {
  # Here we get: 
  # - Long version
  # - Normalized to time 0
  # - Averaged across replicates
  growth_22_avg_norm_long <- process_plate_reader_growth_curves(df = growth_22,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 22,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_26_avg_norm_long <- process_plate_reader_growth_curves(df = growth_26,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 26,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_30_avg_norm_long <- process_plate_reader_growth_curves(df = growth_30,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 30,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_34_avg_norm_long <- process_plate_reader_growth_curves(df = growth_34,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 34,
                                                           normalized_to_t0 = T,
                                                           long_version = T,
                                                           wrong_plate = F)
  # Join results from this into a single long dataframe with all temperatures, to plot from
  growth_long_full_norm <- rbind(rbind(rbind(growth_22_avg_norm_long, growth_26_avg_norm_long), growth_30_avg_norm_long), growth_34_avg_norm_long)
  
  
  # Here we get: 
  # - Wide version
  # - No normalization
  # - Averaged across replicates
  # --> In order to check what was the issue at 30ÂºC!
  growth_22_avg <- process_plate_reader_growth_curves(df = growth_22,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 22,
                                                           normalized_to_t0 = F,
                                                           long_version = F,
                                                           wrong_plate = F)
  
  growth_26_avg <- process_plate_reader_growth_curves(df = growth_26,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 26,
                                                           normalized_to_t0 = F,
                                                           long_version = F,
                                                           wrong_plate = F)
  
  growth_30_avg <- process_plate_reader_growth_curves(df = growth_30,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 30,
                                                           normalized_to_t0 = F,
                                                           long_version = F,
                                                           wrong_plate = F)
  
  growth_34_avg <- process_plate_reader_growth_curves(df = growth_34,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 34,
                                                           normalized_to_t0 = F,
                                                           long_version = F,
                                                           wrong_plate = F)
  
  
  # Here we get: 
  # - Long version
  # - No normalization
  # - Averaged across replicates
  growth_22_avg_long_raw <- process_plate_reader_growth_curves(df = growth_22,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 22,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_26_avg_long_raw <- process_plate_reader_growth_curves(df = growth_26,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 26,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_30_avg_long_raw <- process_plate_reader_growth_curves(df = growth_30,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 30,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)
  
  growth_34_avg_long_raw <- process_plate_reader_growth_curves(df = growth_34,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 34,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)
  # Join results from this into a single long dataframe with all temperatures, to plot from
  growth_long_full_raw <- rbind(rbind(rbind(growth_22_avg_long_raw, growth_26_avg_long_raw), growth_30_avg_long_raw), growth_34_avg_long_raw)
}
```

## 1.3. In the long versions of the datasets, create a new column with the time in hours instead of seconds, quite a bit more understandable
```{r}
growth_long_full_norm <- growth_long_full_norm %>%
  dplyr::mutate(Time_secs = Time,
                Time = Time_secs/3600)

growth_long_full_raw <- growth_long_full_raw %>%
  dplyr::mutate(Time_secs = Time,
                Time = Time_secs/3600)
```




# 2. Initial checks
## 2.1. How stable was the temperature along each of the cultures?
```{r}
# In the first set of measurements, first two temperatures were in one incubator, next two in another one, should check the names and put them somewhere in the graph - similar for the second set, there is a mix of the 2 different incubators/PlateReaders/whatever they are
# They should be in the second sheet of each of these Excel files
# Be careful, I am not doing this from the original data but from the long version, so if I change anything it might get biased (more observations from one temperature than another or whatever)

if (which_try == "first_try") {
  ggplot() +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 26), aes(x = Time, y = Temp_observed), col = "darkgreen", linewidth = 1) +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 30), aes(x = Time, y = Temp_observed), col = "blue", linewidth = 1) +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 34), aes(x = Time, y = Temp_observed), col = "purple", linewidth = 1) +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 37), aes(x = Time, y = Temp_observed), col = "orange", linewidth = 1) +
  geom_hline(yintercept = 26, col = "black") +
  geom_hline(yintercept = 30, col = "black") +
  geom_hline(yintercept = 34, col = "black") +
  geom_hline(yintercept = 37, col = "black") +
  theme_light() +
  xlab("Time (h)") +
  ylab("Temperature") +
  labs(title = "How well did the incubators maintain their assigned temperature?",
       subtitle = try_for_plots)
} else 
  if (which_try == "second_try") {
  ggplot() +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 22), aes(x = Time, y = Temp_observed), col = "orange", linewidth = 1) +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 26), aes(x = Time, y = Temp_observed), col = "darkgreen", linewidth = 1) +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 30), aes(x = Time, y = Temp_observed), col = "blue", linewidth = 1) +
  geom_line(data = subset(growth_long_full_norm, Temp_set == 34), aes(x = Time, y = Temp_observed), col = "purple", linewidth = 1) +
  geom_hline(yintercept = 22, col = "black") +
  geom_hline(yintercept = 26, col = "black") +
  geom_hline(yintercept = 30, col = "black") +
  geom_hline(yintercept = 34, col = "black") +
  theme_light() +
  xlab("Time (h)") +
  ylab("Temperature") +
  labs(title = "How well did the incubators maintain their assigned temperature?",
       subtitle = try_for_plots)
}

```

## 2.2. How consistent is the initial OD across all samples and temperatures - to see if normalizing to it is reasonable
I would say that at least for the second round it does look pretty good?? All initial ODs are within a range of 0.025 OD, seems reasonable to me. 34ÂºC does look generally a bit higher than the other temperatures. 
```{r}
ggplot(data = subset(growth_long_full_raw, Time == 0), 
       aes(x = Strain, y = OD_raw, col = as.factor(Temp_set))) +
  geom_point() +
  theme_light() +
  labs(title = "ODs at time = 0, for each strain at each temperature",
       subtitle = try_for_plots,
       col = "Temperature") +
  ylab("OD") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
```

## 2.3. Check what is wrong with T = 30ÂºC in the second try of the pilot - no idea, but it is wrong lol, Wenxi is going to repeat it, I think I can delete this
```{r}
#t_26_just_data <- growth_26_avg %>%
#  dplyr::select(-c(Time, Temperature))
#t_30_just_data <- growth_30_avg %>%
#  dplyr::select(-c(Time, Temperature))
#t_34_just_data <- growth_34_avg %>%
#  dplyr::select(-c(Time, Temperature))
#
#
#hist(as.matrix(t_26_just_data))
#hist(as.matrix(t_30_just_data))
#hist(as.matrix(t_34_just_data))
```



# 3. Main growth curves
## 3.1. With the raw ODs
```{r}
plot_list <- list()
strains <- sort(unique(sample_ids$Sys.Name))
for (i in 1:length(strains)) {
  strain <- strains[i]
  strain_type <- sample_ids$Status[sample_ids$Sys.Name == strain][1]
  gene_name <- sample_ids$Gene.name[sample_ids$Sys.Name == strain][1]
  temp <- growth_long_full_raw %>%
    dplyr::filter(Strain == strain)
  if (strain_type == "Synthetase") {
    lethal_temp <- sample_ids$Lethal_temp[sample_ids$Sys.Name == strain][1]
    plot_list[[i]] <- ggplot(data = temp, aes(x = Time, y = OD_raw, group = as.factor(Temp_set), col = as.factor(Temp_set))) +
    geom_line(linewidth = 1) +
    theme_light() +
    xlab("Time (h)") +
    ylab("OD 600") +
    labs(title = glue("{strain} - {strain_type}"),
         subtitle = glue("{gene_name} - {lethal_temp}"),
         col = "Temperature (ÂºC)") +
      coord_cartesian(ylim = c(0, 1.25))
  } else 
  if (strain_type == "Lethal") {
    lethal_temp <- sample_ids$Lethal_temp[sample_ids$Sys.Name == strain][1]
    plot_list[[i]] <- ggplot(data = temp, aes(x = Time, y = OD_raw, group = as.factor(Temp_set), col = as.factor(Temp_set))) +
    geom_line(linewidth = 1) +
    theme_light() +
    xlab("Time (h)") +
    ylab("OD 600") +
    labs(title = glue("{strain} - {strain_type}"),
         subtitle = glue("{lethal_temp}"),
         col = "Temperature (ÂºC)") +
      coord_cartesian(ylim = c(0, 1.25))
  }
  else
  {
    plot_list[[i]] <- ggplot(data = temp, aes(x = Time, y = OD_raw, group = as.factor(Temp_set), col = as.factor(Temp_set))) +
    geom_line(linewidth = 1) +
    theme_light() +
    xlab("Time (h)") +
    ylab("OD 600") +
    labs(title = glue("{strain} - {strain_type}"),
         col = "Temperature (ÂºC)") +
      coord_cartesian(ylim = c(0, 1.25))
  }
}

## List of ggplot objects has to be turned to list of grobs for grid.arrange to work
my_grobs <- lapply(plot_list, ggplotGrob)

## Print to PDF
output_file = paste(base_dir, "TSSC/Output/Plots/pilot/growth_curves/gcs_trna_synthetases_raw_", which_try, ".pdf", sep = "")
ggsave(
   filename = output_file, 
   plot = marrangeGrob(my_grobs, nrow=3, ncol=3), 
   width = 15, height = 9
)

for (i in 1:length(plot_list)) {
  print(plot_list[[i]])
}
```

## 3.2. With the ODs normalized to time = 0
```{r}
plot_list <- list()
temp <- sample_ids %>%
  dplyr::arrange(Status)
strains <- unique(temp$Sys.Name)
for (i in 1:length(strains)) {
  strain <- strains[i]
  strain_type <- sample_ids$Status[sample_ids$Sys.Name == strain][1]
  gene_name <- sample_ids$Gene.name[sample_ids$Sys.Name == strain][1]
  temp <- growth_long_full_norm %>%
    dplyr::filter(Strain == strain)
  if (strain_type == "Synthetase") {
    lethal_temp <- sample_ids$Lethal_temp[sample_ids$Sys.Name == strain][1]
    plot_list[[i]] <- ggplot(data = temp, aes(x = Time, y = OD_norm, group = as.factor(Temp_set), col = as.factor(Temp_set))) +
    geom_line(linewidth = 1) +
    theme_light() +
    xlab("Time (h)") +
    ylab("Ratio to OD at T0") +
    labs(title = glue("{strain} - {strain_type}"),
         subtitle = glue("{gene_name} - {lethal_temp}"),
         col = "Temperature (ÂºC)") +
      coord_cartesian(ylim = c(0, 12.5))
  } else 
  if (strain_type == "Lethal") {
    lethal_temp <- sample_ids$Lethal_temp[sample_ids$Sys.Name == strain][1]
    plot_list[[i]] <- ggplot(data = temp, aes(x = Time, y = OD_norm, group = as.factor(Temp_set), col = as.factor(Temp_set))) +
    geom_line(linewidth = 1) +
    theme_light() +
    xlab("Time (h)") +
    ylab("OD 600") +
    labs(title = glue("{strain} - {strain_type}"),
         subtitle = glue("{lethal_temp}"),
         col = "Temperature (ÂºC)") +
      coord_cartesian(ylim = c(0, 12.5))
  }
  else
  {
    plot_list[[i]] <- ggplot(data = temp, aes(x = Time, y = OD_norm, group = as.factor(Temp_set), col = as.factor(Temp_set))) +
    geom_line(linewidth = 1) +
    theme_light() +
    xlab("Time (h)") +
    ylab("Ratio to OD at T0") +
    labs(title = glue("{strain} - {strain_type}"),
         col = "Temperature (ÂºC)") +
      coord_cartesian(ylim = c(0, 12.5))
  }
}

## List of ggplot objects has to be turned to list of grobs for grid.arrange to work
my_grobs <- lapply(plot_list, ggplotGrob)

## Print to PDF
output_file = paste(base_dir, "TSSC/Output/Plots/pilot/growth_curves/gcs_trna_synthetases_norm_to_t0_", which_try, ".pdf", sep = "")
ggsave(
   filename = output_file, 
   plot = marrangeGrob(my_grobs, nrow=3, ncol=3), 
   width = 15, height = 9
)

for (i in 1:length(plot_list)) {
  print(plot_list[[i]])
}
```



# 4. Evaluate time at which OD = 0.5 is reached per temperature and per strain - don't think this makes a lot of sense or is necessary tbh, could probably delete it
```{r}
growth_long_full_raw_OD_05 <- growth_long_full_raw %>%
  dplyr::filter(OD_raw > 0.45 & OD_raw < 0.55)

plot_list <- list()
strains <- sort(sample_ids$Sys.Name[sample_ids$Status == "Synthetase"|sample_ids$Status == "Control"|sample_ids$Status == "BY4741"])

for (i in 1:length(strains)) {
  strain <- strains[[i]]
  temp <- growth_long_full_raw_OD_05 %>%
    dplyr::filter(Strain == strain)
  temp$Temp_set <- as.factor(temp$Temp_set)
  plot_list[[i]] <- ggplot(data = temp, aes(x = Temp_set, y = Time, group = Temp_set)) +
    geom_boxplot(aes(fill = Temp_set)) +
    theme_light() +
    labs(title = glue("{strain}"),
         fill = "Temperature") +
    xlab("Temperature") +
    ylab("Time (h)")
}


## List of ggplot objects has to be turned to list of grobs for grid.arrange to work
my_grobs <- lapply(plot_list, ggplotGrob)

## Print to PDF
output_file = paste(base_dir, "TSSC/Output/Plots/pilot/growth_curves/gcs_trna_synthetases_raw_OD_05_boxplots_", which_try, ".pdf", sep = "")
ggsave(
   filename = output_file, 
   plot = marrangeGrob(my_grobs, nrow=3, ncol=3), 
   width = 15, height = 9
)

for (i in 1:length(plot_list)) {
  print(plot_list[[i]])
}
```





# 5. Check which references do the tRNA synthetase ts alleles come from, so I can maybe have a look at them and get more information on how they work
```{r}
# Synthetases
synthetases <- fread(paste(base_dir, "tRNA_KOs/Data/basic/trna_synthetases.csv", sep=""))

# Supplementary table from Li et al., 2011 (contains the information of the references)
li_et_al <- fread(paste(base_dir, "TSSC/Data/Articles/Li_et_al_2011/Li_et_al_S1.csv", sep="")) %>%
  dplyr::select(Sys.Name, Reference.or.Source, ts.on.SC) %>%
  dplyr::rename(Gene.secondaryIdentifier = Sys.Name) %>%
  dplyr::filter(Gene.secondaryIdentifier %in% synthetases$Gene.secondaryIdentifier)

print(li_et_al$Reference.or.Source)
```
--> Only 5 of them were present in the original library :(



# 6. Sample layout for proteomics experiment
## 6.1. Prepare the sample layout
```{r}
# Grab the strains we want to use for the proteomics experiment: the tRNA synthetase ts strains, and 2 of the replicates of the control strain
strains_for_proteomics <- sample_ids %>%
  dplyr::filter(Status == "Synthetase"|Sys.Name == "Control_1"|Sys.Name == "Control_2")

# Add columns with the average OD of each strain at each temperature
temperatures <- unique(growth_long_full_raw$Temp_set)
strains <- strains_for_proteomics$Sys.Name

# Iterate over temperatures
for (i in 1:length(temperatures)) {
  temperature <- temperatures[i]
  avg_ods_this_temp_20 <- c()
  avg_ods_this_temp_48 <- c()
  
  # For each temperature, iterates over strains, and get the average OD for each of them (at this temperature - at 20h and 48h)
  for (j in 1:length(strains)) {
    strain <- strains[j]
    temp_20 <- growth_long_full_raw %>%
      dplyr::filter(Strain == strain & Temp_set == temperature & Time > 19.5 & Time < 20.5)     # Time is not always exactly 20h
    temp_48 <- growth_long_full_raw %>%
      dplyr::filter(Strain == strain & Temp_set == temperature & Time > 47.5 & Time < 48.5)     # Time is not always exactly 48h
    avg_ods_this_temp_20 <- c(avg_ods_this_temp_20, mean(temp_20$OD_raw))
    avg_ods_this_temp_48 <- c(avg_ods_this_temp_48, mean(temp_48$OD_raw))
  }
  
  # Add the vectors containing the collected average ODs at 20h and 48h at this temperature to the output dataframe
  strains_for_proteomics[[paste("OD_at_", temperature, "_20h", sep="")]] <- avg_ods_this_temp_20
  strains_for_proteomics[[paste("OD_at_", temperature, "_48h", sep="")]] <- avg_ods_this_temp_48
}

# Add two columns which summarize whether this strain is viable (OD > 0.2) at 48h - at 30ÂºC and 34ÂºC
strains_for_proteomics <- strains_for_proteomics %>%
  dplyr::mutate(Viable_at_30 = case_when(OD_at_30_48h > 0.2 ~ "Yes",
                                         TRUE ~ "No"),
                Viable_at_34 = case_when(OD_at_34_48h > 0.2 ~ "Yes",
                                         TRUE ~ "No"))


# Get 2 versions of this dataframe: one for the first 3 rows (to be harvested at ~20h), another one for the last 3 rows (to be harvested at ~48h)
# Each of these is first triplicated (since each row in the dataframe represents a strain, and we'll have 3 replicates of each), then each of the dataframes is randomized separately, so the sample layout will not be the same across the first 3 rows and across the last 3 - I hope this is feasible with the Rotor or Pixl or whatever it is?? If I give it the location of each of them in the old plate?? So that we can randomize the samples, which I think is good practice, but obviously so that Wenxi doesn't have to do it manually
# Anyway, after randomizing the rows in each dataframe, I replicate this three times 
strains_for_proteomics_randomized_1 <- slice(strains_for_proteomics, rep(1:nrow(strains_for_proteomics), 3))                            # Get triplicates of the rows
strains_for_proteomics_randomized_1 <- strains_for_proteomics_randomized_1[sample(nrow(strains_for_proteomics_randomized_1)),] %>%      # Randomize row order
  dplyr::mutate(Column = rep(seq(from = 1, to = 12, by = 1), 3),                                                                        # Assign columns
                Row = c(rep("A", 12), rep("B", 12), rep("C", 12))) %>%                                                                  # Assign rows
  add_well_IDs(column_numbers_col = "Column",
               row_letters_col = "Row")

strains_for_proteomics_randomized_2 <- slice(strains_for_proteomics, rep(1:nrow(strains_for_proteomics), 3))                            # Get triplicates of the rows
strains_for_proteomics_randomized_2 <- strains_for_proteomics_randomized_2[sample(nrow(strains_for_proteomics_randomized_2)),] %>%      # Randomize row order
  dplyr::mutate(Column = rep(seq(from = 1, to = 12, by = 1), 3),                                                                        # Assign columns
                Row = c(rep("E", 12), rep("F", 12), rep("G", 12))) %>%                                                                  # Assign rows
  add_well_IDs(column_numbers_col = "Column",
               row_letters_col = "Row")

# Join the 2 so we get the whole layout for the plate
strains_for_proteomics_final <- rbind(strains_for_proteomics_randomized_1, strains_for_proteomics_randomized_2)
```


## 6.2. Draw 96-well plate with the layout produced above
```{r}
# Color by strain
raw_map(data = strains_for_proteomics_final$Sys.Name,
        well = strains_for_proteomics_final$Well_ID,
        plate = 96) +
  ggtitle("Sample layout for proteomics experiment") +
  theme_dark() +
  theme(axis.text=element_text(size=6),
        legend.key.size = unit(0.5, 'cm'),          # change legend key size
        legend.key.height = unit(0.5, 'cm'),        # change legend key height
        legend.key.width = unit(0.5, 'cm')) +       # change legend key width
  guides(fill = guide_legend(ncol = 2)) +
  labs(fill = "Strain")

# Color by strain type
raw_map(data = strains_for_proteomics_final$Status,
        well = strains_for_proteomics_final$Well_ID,
        plate = 96) +
  theme_dark() +
  theme(axis.text=element_text(size=6),
        legend.key.size = unit(0.5, 'cm'),          # change legend key size
        legend.key.height = unit(0.5, 'cm'),        # change legend key height
        legend.key.width = unit(0.5, 'cm')) +       # change legend key width)
  labs(title = "Sample layout for proteomics experiment",
       fill = "Strain type")

# Color by viable at 30ÂºC
raw_map(data = strains_for_proteomics_final$Viable_at_30,
        well = strains_for_proteomics_final$Well_ID,
        plate = 96) +
  theme_dark() +
  theme(axis.text=element_text(size=6),
        legend.key.size = unit(0.5, 'cm'),          # change legend key size
        legend.key.height = unit(0.5, 'cm'),        # change legend key height
        legend.key.width = unit(0.5, 'cm')) +       # change legend key width)
  labs(title = "Sample layout for proteomics experiment",
       fill = "Viable at 30ÂºC")

# Color by viable at 34ÂºC
raw_map(data = strains_for_proteomics_final$Viable_at_34,
        well = strains_for_proteomics_final$Well_ID,
        plate = 96) +
  theme_dark() +
  theme(axis.text=element_text(size=6),
        legend.key.size = unit(0.5, 'cm'),          # change legend key size
        legend.key.height = unit(0.5, 'cm'),        # change legend key height
        legend.key.width = unit(0.5, 'cm')) +       # change legend key width)
  labs(title = "Sample layout for proteomics experiment",
       fill = "Viable at 34ÂºC")
```

## 6.3. Get a drawing of the plate of the growth curves
For comparison in the presentation, to show how all replicates were together
```{r}
# Come up with the full sample IDs of the plate
sample_ids_new_2 <- create_full_sample_ids_for_a_plate(df = growth_26,
                                                           sample_ids = sample_ids, 
                                                           replicates = list(c("A", "B", "C"), c("E", "F", "G")),
                                                           empty_rows = c("D", "H"),
                                                           fixed_temp = 26,
                                                           normalized_to_t0 = F,
                                                           long_version = T,
                                                           wrong_plate = F)

# Fix well IDs
sample_ids_new_2 <- fix_well_ids_with_no_0s(sample_ids_new_2, "GC96")

# Color by viable at 34ÂºC
raw_map(data = sample_ids_new_2$Sys.Name,
        well = sample_ids_new_2$GC96,
        plate = 96) +
  theme_dark() +
  theme(axis.text=element_text(size=6),
        legend.key.size = unit(0.5, 'cm'),          # change legend key size
        legend.key.height = unit(0.5, 'cm'),        # change legend key height
        legend.key.width = unit(0.5, 'cm')) +       # change legend key width)
  labs(title = "Sample layout for growth curves",
       fill = "Strains")
```




























